docs/proptest.md

```diff
@@ -0,0 +1,104 @@
+repository](https://github.com/ComposableFi/composable). To be more precise, we
+will only use the [Vault
+pallet](https://github.com/ComposableFi/composable/tree/main/frame/vault).
+However, to generalize these ideas over arbitrary projects, we will try to stay a bit abstract.
+
+## Determining a property
+
+Before being able to determine any property, we have to understand the code we
+are working on. The Vault allows strategies to operate on users funds. Users are
+able to deposit an arbitrary amount and get LP tokens representing their share
+of the vault. They later can withdraw using these LPs. Usually, this kind of
+protocol's purpose is to have a better capital efficiency and optimize APY over
+this capital.
+
+One of the easiest properties here is the fact that `withdraw . deposit =
+identity`. This means that if a user **deposit** and then **withdraw**, the vault
+should be unchanged and the user balance should be unchanged. Note that the
+**amount** is not part of the equation, which allows us to generate any
+amount and assert this property.
+
+## Generating random inputs
+
+Recall the `withdraw . deposit = identity` which we just found. We said that the
+amount was not part of the property. Interestingly, we no longer need to have
+hardcoded values as long as they are not part of the equation.
+
+Usually, the property we want to check is valid for the whole input domain. But
+this is not always the case (it should probably be enforced but in practice, we
+encounter a lot of functions like this). For our example, any **amount** is
+valid for the property to hold, the generating function will be simple.
+This example we are going to work on is extracted from [the Vault tests](https://github.com/ComposableFi/composable/blob/71af1080f3c89233e289e41b3a249d0966aa1147/frame/vault/src/tests.rs#L271).
+
+Let's assume that the `deposit` and `withdraw` functions are using a `type
+Balance = u128` as input, using proptest [`prop_compose!`](https://altsysrq.github.io/rustdoc/proptest/1.0.0/proptest/macro.prop_compose.html) macro, we can create a
+generator function like:
+```rust
+prop_compose! {
+	fn random_valid_amount()(x in 0..Balance::MAX) -> Balance {
+		x
+	}
+}
+```
+
+## Using generator function in tests inputs
+
+The `random_valid_amount` generator we just defined can directly be used to
+generate random inputs for functions defined inside proptest [`proptest!`](https://altsysrq.github.io/rustdoc/proptest/1.0.0/proptest/macro.proptest.html) macro:
+```rust
+proptest! {
+    // Declare that every test within this macro should be executed 10_000 times (new random input each time) to pass.
+	#![proptest_config(ProptestConfig::with_cases(10_000))]
+    
+	#[test]
+	fn test_withdraw_deposit_identity(amount in random_valid_amount()) {
+       // TODO: check our property using the random `amount`
+       // recall that this function we be executed 10_000 times when running `cargo test`
```

Comment:
can use a tab before those lines instead of spaces to align with the lines below.